libtransmission/settings.h

```diff
@@ -32,26 +25,52 @@ class Settings
     [[nodiscard]] tr_variant save() const;
 
 protected:
-    using field_key_type = tr_quark;
-    using field_mapped_type = std::variant<
-        bool*,
-        double*,
-        size_t*,
-        std::chrono::milliseconds*,
-        std::string*,
-        tr_encryption_mode*,
-        tr_log_level*,
-        tr_mode_t*,
-        tr_open_files::Preallocation*,
-        tr_port*,
-        tr_preferred_transport*,
-        tr_tos_t*,
-        tr_verify_added_mode*>;
-    using field_value_type = std::pair<field_key_type const, field_mapped_type>;
-    using Fields = std::vector<field_value_type>;
-
-    Settings() = default;
+    Settings();
+
+    // convert from tr_variant to T
+    template<typename T>
+    using Load = bool (*)(tr_variant const& src, T* tgt);
+
+    // convert from T to tr_variant
+    template<typename T>
+    using Save = tr_variant (*)(T const& src);
+
+    template<typename T>
+    void add_type_handler(Load<T> load, Save<T> save)
+    {
+        auto const key = std::type_index(typeid(T*));
+
+        // wrap load + save with void* wrappers so that
+        // they can be stored in the save_ and load_ maps
+        load_.insert_or_assign(key, [load](tr_variant const& src, void* tgt) { return load(src, static_cast<T*>(tgt)); });
+        save_.insert_or_assign(key, [save](void const* src) { return save(*static_cast<T const*>(src)); });
+    }
+
+    class Field
+    {
+    public:
+        template<typename T>
+        Field(tr_quark key_in, T* ptr_in)
+            : key{ key_in }
+            , type{ typeid(T*) }
+            , ptr{ ptr_in }
+        {
+        }
+
+    private:
+        friend Settings;
+
+        tr_quark key;
+        std::type_info const& type;
+        void* ptr;
```

Comment:
Maybe the fields of the `Field` class are better off being public.

- `Field` itself is only exposed to `Settings` and its subclasses.
- The fields are only directly accessed, and never operated on by any class methods.

Reply:
done in c2c9066240420785c996e2f75e93adcd2e0e1c2c